File: strategies/tradepro_strategies/quant_engine/monte_carlo.py
"""Block-bootstrap Monte Carlo simulation.

Draws random blocks of consecutive daily returns (block_size=21 ≈ 1
trading month) and assembles them into synthetic equity paths. This
preserves the autocorrelation and fat-tail structure of actual returns
better than IID sampling.

Each path starts at `initial` and compounds the drawn returns. The
summary dictionary reports the distribution of final portfolio values
and maximum drawdowns across all simulated paths.
"""
from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np
import pandas as pd

from .portfolio_metrics import max_drawdown as _max_dd, cagr as _cagr


@dataclass
class MonteCarloResult:
    """Output from a Monte Carlo simulation run."""
    paths: np.ndarray          # shape (n_sims, n_days+1) — equity paths
    summary: dict
    n_sims: int
    years: int


class MonteCarloSimulator:
    """Block-bootstrap Monte Carlo simulator.

    Parameters
    ----------
    returns : pd.Series
        Daily returns (used as the empirical distribution to sample from).
    block_size : int
        Length of each bootstrap block in trading days. Default 21.
    seed : int or None
        Random seed for reproducibility.
    """

    def __init__(
        self,
        returns: pd.Series,
        block_size: int = 21,
        seed: int | None = None,
    ) -> None:
        self.returns = returns.dropna().values
        self.block_size = block_size
        self.seed = seed

    def run(
        self,
        initial: float = 10_000.0,
        years: int = 10,
        n_sims: int = 1000,
        periods_per_year: int = 252,
    ) -> MonteCarloResult:
        """Run the simulation and return a MonteCarloResult.

        Parameters
        ----------
        initial : float
            Starting portfolio value.
        years : int
            Simulation horizon in years.
        n_sims : int
            Number of simulation paths.
        periods_per_year : int
            Trading days per year (252 for daily).
        """
        rng = np.random.default_rng(self.seed)
        n_days = int(years * periods_per_year)
        src = self.returns
        n_src = len(src)
        bs = self.block_size

        if n_src == 0:
            raise ValueError("returns series is empty")

        # Build paths array: shape (n_sims, n_days+1)
        paths = np.empty((n_sims, n_days + 1))
        paths[:, 0] = initial

        for sim_idx in range(n_sims):
            daily = np.empty(n_days)
            filled = 0
            while filled < n_days:
                # Draw a random starting position for the block
                start = int(rng.integers(0, max(1, n_src - bs + 1)))
                block = src[start: start + bs]
                take = min(len(block), n_days - filled)
                daily[filled: filled + take] = block[:take]
                filled += take
            # Compound
            paths[sim_idx, 1:] = initial * np.cumprod(1.0 + daily)

        final_values = paths[:, -1]
        summary = self._compute_summary(paths, final_values, initial, years, periods_per_year)

        return MonteCarloResult(
            paths=paths,
            summary=summary,
            n_sims=n_sims,
            years=years,
        )

    def _compute_summary(
        self,
        paths: np.ndarray,
        final_values: np.ndarray,
        initial: float,
        years: int,
        periods_per_year: int,
    ) -> dict:
        pct_labels = [5, 10, 25, 50, 75, 90, 95]
        percentiles_vals = np.percentile(final_values, pct_labels)

        def _cagr_from_fv(fv: float) -> float:
            if initial <= 0 or years <= 0:
                return 0.0
            return float((fv / initial) ** (1.0 / years) - 1.0) * 100.0

        percentiles = {}
        for pct, fv in zip(pct_labels, percentiles_vals):
            percentiles[f"p{pct}"] = {
                "final_value": round(float(fv), 2),
                "multiple": round(float(fv / initial), 4) if initial > 0 else 0.0,
                "cagr_pct": round(_cagr_from_fv(fv), 4),
            }

        # Max drawdown distribution across paths
        mdd_per_path = []
        for i in range(paths.shape[0]):
            equity = pd.Series(paths[i])
            mdd_per_path.append(_max_dd(equity) * 100.0)  # as %
        mdd_arr = np.array(mdd_per_path)

        max_dd_pct = {
            f"p{p}": round(float(np.percentile(mdd_arr, p)), 4)
            for p in [5, 25, 50, 75, 95]
        }

        return {
            "percentiles": percentiles,
            "mean_final": round(float(final_values.mean()), 2),
            "p_lose_money": round(float((final_values < initial).mean()), 4),
            "p_double": round(float((final_values >= initial * 2).mean()), 4),
            "p_5x": round(float((final_values >= initial * 5).mean()), 4),
            "max_dd_pct": max_dd_pct,
        }

File: frontend/src/util/monteCarlo.ts
/**
 * Client-side block-bootstrap Monte Carlo — mirrors the Python quant engine
 * (`quant_engine/monte_carlo.py`) so per-symbol projections can run in the
 * browser from candle returns without a backend round-trip. The portfolio /
 * strategy MC is still computed server-side (richer, ensemble returns); this
 * produces the SAME data shape (`MonteCarloData`) so one renderer serves both.
 *
 * Method: draw random BLOCKS of consecutive daily returns (default 21 ≈ 1
 * trading month) and stitch them into synthetic equity paths. Blocks (not IID
 * single days) preserve autocorrelation + fat tails. Each path compounds from
 * `initial`. We report the distribution of final values + max drawdowns, and a
 * down-sampled percentile fan over time.
 *
 * Honest-by-construction: returns null when there isn't enough history to
 * sample (callers render "not enough data" rather than a fabricated cone).
 */

/** Percentile-fan + summary payload — matches the server MC shape so the same
 *  chips + fan chart render server (portfolio) and client (per-symbol) results. */
export interface MonteCarloData {
  fan_chart: {
    years_axis: number[];
    q05: number[];
    q25: number[];
    q50: number[];
    q75: number[];
    q95: number[];
  };
  summary: {
    percentiles: Record<string, { final_value: number; multiple: number; cagr_pct: number }>;
    p_double: number;
    p_5x: number;
    p_lose_money: number;
    max_dd_pct: { p50: number };
  };
  n_sims: number;
  years: number;
  initial: number;
  source: "client" | "server";
}

export interface MonteCarloOpts {
  years?: number;
  nSims?: number;
  initial?: number;
  blockSize?: number;
  periodsPerYear?: number;
  seed?: number;
  /** ~number of x-axis points in the fan (paths are sampled to keep it light). */
  fanPoints?: number;
}

/** Deterministic RNG (mulberry32) — same seed ⇒ same projection (reproducible). */
function mulberry32(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a |= 0;
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function percentile(sorted: number[], p: number): number {
  if (sorted.length === 0) return 0;
  const idx = (p / 100) * (sorted.length - 1);
  const lo = Math.floor(idx);
  const hi = Math.ceil(idx);
  if (lo === hi) return sorted[lo];
  return sorted[lo] + (sorted[hi] - sorted[lo]) * (idx - lo);
}

/** Daily simple returns from a close series (drops non-finite + the first NaN). */
export function returnsFromCloses(closes: Array<number | null | undefined>): number[] {
  const c = closes.filter((v): v is number => typeof v === "number" && Number.isFinite(v) && v > 0);
  const out: number[] = [];
  for (let i = 1; i < c.length; i++) out.push(c[i] / c[i - 1] - 1);
  return out;
}

/**
 * Run the block-bootstrap. Returns null when `returns` has fewer than
 * `blockSize` observations (can't draw a single block honestly).
 */
export function runBlockBootstrapMC(
  returns: number[],
  opts: MonteCarloOpts = {},
): MonteCarloData | null {
  const years = opts.years ?? 5;
  const nSims = opts.nSims ?? 500;
  const initial = opts.initial ?? 10_000;
  const blockSize = opts.blockSize ?? 21;
  const ppy = opts.periodsPerYear ?? 252;
  const seed = opts.seed ?? 42;
  const fanPoints = Math.max(2, opts.fanPoints ?? 120);

  const src = returns.filter((r) => Number.isFinite(r));
  const nSrc = src.length;
  if (nSrc < blockSize) return null;

  const nDays = Math.max(1, Math.round(years * ppy));
  const rng = mulberry32(seed);

  // Sample indices for the fan (down-sampled time axis, always incl. day 0 + last).
  const step = Math.max(1, Math.floor(nDays / (fanPoints - 1)));
  const sampleIdx: number[] = [];
  for (let d = 0; d <= nDays; d += step) sampleIdx.push(d);
  if (sampleIdx[sampleIdx.length - 1] !== nDays) sampleIdx.push(nDays);

  // sampledCols[k] = array of equity values across sims at sampleIdx[k].
  const sampledCols: number[][] = sampleIdx.map(() => new Array<number>(nSims));
  const finals = new Array<number>(nSims);
  const maxDDs = new Array<number>(nSims);

  for (let s = 0; s < nSims; s++) {
    let equity = initial;
    let peak = initial;
    let maxDD = 0; // most-negative (equity-peak)/peak
    let day = 0;
    let sampleK = 0;
    // record day 0
    if (sampleIdx[sampleK] === 0) sampledCols[sampleK++][s] = equity;

    while (day < nDays) {
      const start = Math.floor(rng() * Math.max(1, nSrc - blockSize + 1));
      const take = Math.min(blockSize, nDays - day);
      for (let b = 0; b < take; b++) {
        equity *= 1 + src[start + b];
        day++;
        if (equity > peak) peak = equity;
        const dd = peak > 0 ? (equity - peak) / peak : 0;
        if (dd < maxDD) maxDD = dd;
        while (sampleK < sampleIdx.length && sampleIdx[sampleK] === day) {
          sampledCols[sampleK++][s] = equity;
        }
      }
    }
    finals[s] = equity;
    maxDDs[s] = maxDD * 100; // as percent (negative)
  }

  // Per-step percentiles for the fan.
  const q05: number[] = [], q25: number[] = [], q50: number[] = [], q75: number[] = [], q95: number[] = [];
  for (let k = 0; k < sampleIdx.length; k++) {
    const col = sampledCols[k].slice().sort((a, b) => a - b);
    q05.push(percentile(col, 5));
    q25.push(percentile(col, 25));
    q50.push(percentile(col, 50));
    q75.push(percentile(col, 75));
    q95.push(percentile(col, 95));
  }
  const years_axis = sampleIdx.map((d) => d / ppy);

  // Final-value percentiles + headline probabilities.
  const finalsSorted = finals.slice().sort((a, b) => a - b);
  const pctLabels = [5, 10, 25, 50, 75, 90, 95];
  const percentiles: MonteCarloData["summary"]["percentiles"] = {};
  for (const p of pctLabels) {
    const fv = percentile(finalsSorted, p);
    percentiles[`p${p}`] = {
      final_value: Math.round(fv * 100) / 100,
      multiple: initial > 0 ? Math.round((fv / initial) * 1e4) / 1e4 : 0,
      cagr_pct: initial > 0 && years > 0 ? (Math.pow(fv / initial, 1 / years) - 1) * 100 : 0,
    };
  }
  const frac = (pred: (v: number) => boolean) => finals.reduce((n, v) => n + (pred(v) ? 1 : 0), 0) / nSims;
  const ddSorted = maxDDs.slice().sort((a, b) => a - b);

  return {
    fan_chart: { years_axis, q05, q25, q50, q75, q95 },
    summary: {
      percentiles,
      p_double: frac((v) => v >= 2 * initial),
      p_5x: frac((v) => v >= 5 * initial),
      p_lose_money: frac((v) => v < initial),
      max_dd_pct: { p50: percentile(ddSorted, 50) },
    },
    n_sims: nSims,
    years,
    initial,
    source: "client",
  };
}
